File: java/structural_pattern/composite/Folder.java
package composite;

import java.util.ArrayList;
import java.util.List;

public class Folder implements FileComponent {
    private String name;
    private List<FileComponent> children = new ArrayList<>();

    public Folder(String name) {
        this.name = name;
    }

    public void add(FileComponent component) {
        children.add(component);
    }

    public void remove(FileComponent component) {
        children.remove(component);
    }

    @Override
    public void showDetails(int indent) {
        System.out.println(" ".repeat(indent) + "📁 Folder: " + name);
        for (FileComponent child : children) {
            child.showDetails(indent + 1);
        }
    }

    @Override
    public int getSize() {
        int totalSize = 0;
        for (FileComponent child : children) {
            totalSize += child.getSize();
        }
    
        return totalSize;
    }

    @Override
    public String getName() {
        return name;
    }

    public List<FileComponent> getChildren() {
        return children;
    }
}

File: typescript/structural_pattern/src/composite/Folder.ts
import FileComponent from "./FileComponent";

class Folder implements FileComponent {
    private name: string;
    private children: FileComponent[] = [];

    public constructor(name: string) {
        this.name = name;
    }

    public add(component: FileComponent): void {
        this.children.push(component);
    }

    public remove(component: FileComponent): void {
        this.children.map((v) => v !== component);
    }

    public showDetails(indent: number): void {
        console.log(" ".repeat(indent) + "Folder: " + this.name);
        for (let i = 0; i < this.children.length; i++) {
            this.children[i].showDetails(indent + 1);
        }
    }

    public getSize(): number {
        let total = 0;
        for (let i = 0; i < this.children.length; i++) {
            total += this.children[i].getSize();
        }

        return total;
    }

    public getName(): string {
        return this.name;
    }

    public getChildren(): FileComponent[] {
        return this.children;
    }
}

export default Folder;